monero/wallet_client.go

```diff
@@ -221,49 +250,211 @@ func (c *walletClient) WaitForReceipt(req *WaitForReceiptRequest) (*wallet.Trans
 		Address:      string(to),
 	})
+	if err != nil {
+		log.Warnf("Sweep transaction(s) from %s failed, %s", from, err)
+		return nil, fmt.Errorf("sweep_all failed: %w", err)
+	}
+	log.Infof("Sweep transaction started, TX IDs: %s", strings.Join(reqResp.TxHashList, ", "))
+	var transfers []*wallet.Transfer
+	for _, txID := range reqResp.TxHashList {
+		receipt, err := c.waitForReceipt(&waitForReceiptRequest{
+			Ctx:              ctx,
+			TxID:             txID,
+			NumConfirmations: numConfirmations,
+			AccountIdx:       accountIdx,
+		})
+		if err != nil {
+			log.Warnf("Sweep TX ID %s failed waiting for receipt: %s", txID, err)
+			return nil, fmt.Errorf("sweep failed waiting for receipt: %w", err)
+		}
+		log.Infof("Sweep transfer ID=%s of %s XMR (%s XMR fees) completed at height %d",
+			txID,
+			coins.FmtPiconeroAmtAsXMR(receipt.Amount),
+			coins.FmtPiconeroAmtAsXMR(receipt.Fee),
+			receipt.Height,
+		)
+		transfers = append(transfers, receipt)
+	}
+	return transfers, nil
+}
+
+func (c *walletClient) CreateABWalletConf(walletNamePrefix string) *WalletClientConf {
+	walletName := fmt.Sprintf("%s-%s", walletNamePrefix, time.Now().Format(common.TimeFmtNSecs))
+	walletPath := path.Join(path.Dir(c.conf.WalletFilePath), walletName)
+	conf := &WalletClientConf{
+		Env:                 c.conf.Env,
+		WalletFilePath:      walletPath,
+		WalletPassword:      c.conf.WalletPassword,
+		WalletPort:          0,
+		MonerodNodes:        c.conf.MonerodNodes,
+		MoneroWalletRPCPath: c.conf.MoneroWalletRPCPath,
+		LogPath:             c.conf.LogPath,
+	}
+	return conf
+}
+
+func createWalletFromKeys(
+	conf *WalletClientConf,
+	walletRestoreHeight uint64,
+	privateSpendKey *mcrypto.PrivateSpendKey, // nil for a view-only wallet
+	privateViewKey *mcrypto.PrivateViewKey,
+	address mcrypto.Address,
+) (WalletClient, error) {
+	if conf.WalletPort == 0 { // swap wallets need randomized ports, so we expect this to be zero
+		var err error
+		conf.WalletPort, err = getFreeTCPPort()
+		if err != nil {
+			return nil, err
+		}
+	}
+	// should be a one item list, we use the same node that the primary wallet is using
+	monerodNode := conf.MonerodNodes[0]
+
+	proc, err := createWalletRPCService(
+		conf.Env,
+		conf.MoneroWalletRPCPath,
+		conf.WalletPort,
+		path.Dir(conf.WalletFilePath),
+		conf.LogPath,
+		monerodNode,
+	)
+	if err != nil {
+		return nil, err
+	}
+	c := NewThinWalletClient(monerodNode.Host, monerodNode.Port, conf.WalletPort).(*walletClient)
+	c.rpcProcess = proc
+	c.conf = conf
+	err = c.generateFromKeys(
+		privateSpendKey, // nil for a view-only wallet
+		privateViewKey,
+		address,
+		walletRestoreHeight,
+		path.Base(conf.WalletFilePath),
+		conf.WalletPassword,
+	)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+	acctResp, err := c.GetAddress(0)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+	c.walletAddr = mcrypto.Address(acctResp.Address)
+	if c.walletAddr != address {
+		panic("addresses do not match")
```

Comment:
this panic seems risky, is there any way the address passed might not match that generated?

Reply:
I'll add an error level log and return an error. A mismatched address should have been detected long before we get here, but better to be safe and not have the server crash.

Reply:
I don't see this as updated?

Reply:
That code was updated and pushed on the 26th:
https://github.com/AthanorLabs/atomic-swap/blame/dimalinux/reduce-xmr-wallet-contention/monero/wallet_client.go#L412-L417